lang/udf_cache.hh

```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright (C) 2022-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: (AGPL-3.0-or-later and Apache-2.0)
+ */
+
+#pragma once
+
+#include "cql3/functions/function_name.hh"
+#include "utils/overloaded_functor.hh"
+#include "wasmtime.hh"
+#include "lang/wasm.hh"
+#include <seastar/core/shared_ptr.hh>
+#include <seastar/core/scheduling.hh>
+#include "utils/loading_cache.hh"
+#include "utils/hash.hh"
+#include "utils/overloaded_functor.hh"
+#include <chrono>
+#include <cstdint>
+#include <list>
+
+extern logging::logger wasm_logger;
+
+namespace wasm {
+
+struct udf_instance {
+    wasmtime::Store store;
+    wasmtime::Instance instance;
+    wasmtime::Func func;
+    wasmtime::Memory memory;
+};
+
+// For each UDF full name and a scheduling group, we store a future that is ready
+// when no other task uses the same UDF instance. When an instance is retrieved
+// from the cache, the future is replaced with a new one that becomes ready when
+// the instance is returned. To avoid conflicts between tasks, each time an 
+// instance is requested, the future associated with it is removed from the cache
+// and a new one is added. When the current instance user is done with it, the
+// future that was associated with it is becomes ready.
+// This way, an instance stays in the cache as long as it isn't used by any task
+// and if there are tasks that want to use it, they get it immediately after
+// the previous user is finished.
+class udf_cache {
+public:
+    struct stats {
+        uint64_t cache_hits = 0;
+        uint64_t cache_misses = 0;
+        uint64_t cache_blocks = 0;
+    };
+
+    static stats& shard_stats() {
+        static thread_local stats _stats;
+        return _stats;
+    }
+
+private:
+    using cache_key_type = std::pair<seastar::scheduling_group, cql3::functions::function_name>;
+    using cache_entry_type = seastar::future<udf_instance>;
+    struct udf_cache_entry_size {
+        size_t operator()(cache_entry_type& val) {
+            if (!val.available()) {
+                // the instance is currently being used
```

Comment:
Are you using future as a std::optional?

Reply:
And why is an in-use entry in the cache? I expected it would be removed from the cache so new lookups would miss it.

What happens if a lookup finds an in-use entry?

Reply:
I still don't understand completely. Will a still-executing UDF be stored in the cache? Why?

Reply:
I'm worried that even if I understand it today, I won't tomorrow. There are several layers of is-this-instance-usable-or-not, each slightly different.

Reply:
We need to store some kind of entry in the cache even for an in-use UDF, to prevent creation of instances for each concurrent use. It would also be best if this entry allowed notifying the cache user when the entry is no longer in-use.
For this case I'm storing a non-ready future, do you think a semaphore combined with an additional optional instance would be much easier to understand, or is there some other obvious form of an entry that satisfies these conditions?
Perhaps adding a more extensive description in comments would suffice?